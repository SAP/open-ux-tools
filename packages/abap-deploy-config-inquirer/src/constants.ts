export const ABAP_PACKAGE_SEARCH_MAX_RESULTS = 50;
export const DEFAULT_PACKAGE_ABAP = '$tmp';
