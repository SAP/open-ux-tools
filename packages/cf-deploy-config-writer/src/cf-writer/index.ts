export { generateAppConfig } from './app-config';
export { generateRootConfig } from './base-config';
