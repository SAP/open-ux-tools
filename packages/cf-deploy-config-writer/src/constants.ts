import { UI5_DEFAULT } from '@sap-ux/ui5-config';
import { t } from './i18n';

export const WelcomeFile = 'welcomeFile';
export const XSAppFile = 'xs-app.json';
export const XSSecurityFile = 'xs-security.json';
export const MTABuildResult = 'build-result';
export const MTABuildParams = 'build-parameters';
export const MTAFileExtension = 'mta-ext.mtaext';
export const DefaultServiceURL = '${default-url}';
export const ManagedXSUAA = 'managed:xsuaa';
export const HTML5RepoHost = 'html5-apps-repo:app-host';
export const SRV_API = 'srv-api';
export const DefaultMTADestination = 'fiori-default-srv-api';
export const EmptyDestination = 'NONE';
export const ResourceMTADestination = '%s-srv-api';
export const MTAYamlFile = 'mta.yaml';
export const MTADescription = 'Generated by Fiori Tools';
export const RouterModule = 'router';
export const CloudFoundry = 'cf';
export const deployMode = 'deploy_mode';
export const enableParallelDeployments = 'enable-parallel-deployments';
export const CDSExecutable = 'cds';
export const CDSPackage = '@sap/cds-dk';
export const MTAExecutable = 'mta';
export const MTAPackage = 'mta';
export const MTAPackageVersion = '^1.2.27';
export const MTAVersion = '0.0.1';
export const RimrafVersion = '^5.0.5';
export const Rimraf = 'rimraf';
export const MbtPackage = 'mbt';
export const MbtPackageVersion = '^1.2.29';
export const UI5BuilderWebIdePackage = '@sap/ui5-builder-webide-extension';
export const UI5BuilderWebIdePackageVersion = '^1.1.9';
export const UI5TaskZipperPackage = 'ui5-task-zipper';
export const UI5TaskZipperPackageVersion = '^3.1.3';
export const UI5Package = '@ui5/cli';
export const UI5PackageVersion = '^3.9.2';
export const CDSAddMtaParams = ['add', 'mta'];
export const CDSXSUAAService = 'xsuaa';
export const CDSHTML5RepoService = 'html5-repo';
export const CDSConnectivityService = 'connectivity';
export const CDSDestinationService = 'destination';
export const MTAAPIDestination = {
    Name: ResourceMTADestination,
    Type: 'HTTP',
    URL: `~{srv-api/srv-url}`,
    ProxyType: 'Internet',
    Authentication: 'NoAuthentication',
    'HTML5.DynamicDestination': true,
    'HTML5.ForwardAuthToken': true
};
export const UI5Destination = {
    Name: 'ui5',
    Type: 'HTTP',
    URL: UI5_DEFAULT.SAPUI5_CDN,
    ProxyType: 'Internet',
    Authentication: 'NoAuthentication'
};
export const UI5ResourceDestination = {
    'init_data': {
        instance: {
            'existing_destinations_policy': 'update',
            destinations: [
                {
                    Name: 'ui5',
                    Type: 'HTTP',
                    URL: UI5_DEFAULT.SAPUI5_CDN,
                    ProxyType: 'Internet',
                    Authentication: 'NoAuthentication'
                }
            ]
        }
    }
};

export const UI5StandaloneModuleDestination = {
    group: 'destinations',
    properties: {
        forwardAuthToken: false,
        name: 'ui5',
        url: UI5_DEFAULT.SAPUI5_CDN
    }
};

export const DestinationServiceConfig = {
    config: {
        'HTML5Runtime_enabled': true,
        version: '1.0.0',
        ...UI5ResourceDestination
    }
};

export const ServiceAPIRequires = {
    name: SRV_API,
    properties: {
        'srv-url': DefaultServiceURL
    }
};
export const HTMLAppBuildParams = {
    builder: 'custom',
    'build-result': 'dist',
    commands: ['npm install', 'npm run build:cf'],
    'supported-platforms': []
};
export const UI5DeployBuildScript =
    'ui5 build preload --clean-dest --config ui5-deploy.yaml --include-task=generateCachebusterInfo';
export const MTABuildScript = 'rimraf resources mta_archives && mbt build';
export const appDeployMTAScript = (args: string[]): string => {
    const mtaArgs = args.length > 0 ? ` ${args.join(' ')}` : '';
    return `fiori cfDeploy${mtaArgs}`;
};
export const rootDeployMTAScript = (args: string[]): string => {
    const mtaArgs = args.length > 0 ? `${args.join(' ')} ` : '';
    return `cf deploy mta_archives/archive.mtar ${mtaArgs}--retries 1`;
};
export const undeployMTAScript = (mtaId: string): string =>
    `cf undeploy ${mtaId} --delete-services --delete-service-keys --delete-service-brokers`;
const cannotFindBinary = (bin: string, pkg: string): string => t('error.cannotFindBinary', { bin, pkg });
export const CDSBinNotFound = cannotFindBinary(CDSExecutable, CDSPackage);
export const MTABinNotFound = cannotFindBinary(MTAExecutable, MTAPackage);
