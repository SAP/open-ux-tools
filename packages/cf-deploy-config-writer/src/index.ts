export * from './mta-config';
export * from './cf-writer';
export { CFBaseConfig, CFAppConfig, RouterModuleType, ApiHubConfig } from './types';
