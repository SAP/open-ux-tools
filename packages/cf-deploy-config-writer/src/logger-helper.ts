import { ToolsLogger, type Logger } from '@sap-ux/logger';

/**
 * Static logger prevents passing of logger references through all functions, as this is a cross-cutting concern.
 */
export default class LoggerHelper {
    private static _logger: Logger = new ToolsLogger({ logPrefix: '@sap-ux/cf-deploy-config-writer' });

    /**
     * Get the logger.
     *
     * @returns the logger
     */
    public static get logger(): Logger {
        return LoggerHelper._logger;
    }

    /**
     * Set the logger.
     *
     * @param value the logger to set
     */
    public static set logger(value: Logger) {
        LoggerHelper._logger = value;
    }
}
