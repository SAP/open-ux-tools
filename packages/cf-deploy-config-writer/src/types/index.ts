import type { Destination, Authentication } from '@sap-ux/btp-utils';

export type ResourceType =
    | 'xsuaa'
    | 'destination'
    | 'portal'
    | 'connectivity'
    | 'managed:xsuaa'
    | 'html5-apps-repo:app-host'
    | 'html5-apps-repo:app-runtime';
export type ModuleType =
    | 'hdb'
    | 'nodejs'
    | 'java'
    | 'approuter.nodejs'
    | 'com.sap.application.content'
    | 'com.sap.application.content:destination'
    | 'com.sap.application.content:resource'
    | 'html5'
    | 'com.sap.portal.content';
export enum CloudFoundryServiceType {
    Existing = 'org.cloudfoundry.existing-service',
    Managed = 'org.cloudfoundry.managed-service'
}
export type MTADestinationType = Destination & {
    ServiceInstanceName: string;
    ServiceKeyName: string;
    'sap.cloud.service': string;
};
export enum RouterModuleType {
    Standard = 'standard',
    Managed = 'managed'
}
export interface MTABaseConfig {
    mtaId: string;
    mtaPath: string;
    mtaDescription?: string;
    mtaVersion?: string;
}
export interface CFBaseConfig extends MTABaseConfig {
    routerType: RouterModuleType;
    addConnectivityService?: boolean;
    addDestinationService?: boolean;
    abapServiceProvider?: {
        abapServiceName?: string;
        abapService?: string;
    };
}
export interface CFAppConfig {
    appPath: string;
    addManagedRouter?: boolean;
    destination?: string;
    apiHubConfig?: ApiHubConfig;
    addMTADestination?: boolean;
    serviceHost?: string; // Data service host
    lcapMode?: boolean;
}
export interface CFConfig extends CFAppConfig, CFBaseConfig {
    appId: string;
    rootPath: string;
    serviceBase: string;
    capRoot?: string;
    isCap?: boolean;
    servicePath?: string;
    firstServicePathSegment?: string;
    isFullUrlDest?: boolean;
    destinationAuthType?: Authentication;
    cloudServiceName?: string;
    isMtaRoot?: boolean;
}
export const enum ApiHubType {
    apiHub = 'API_HUB',
    apiHubEnterprise = 'API_HUB_ENTERPRISE'
}
export interface ApiHubConfig {
    apiHubKey: string;
    apiHubType: ApiHubType;
}
export interface XSAppRoute {
    source?: string;
    target?: string;
    destination?: string;
    csrfProtection?: boolean;
    scope?: string;
    service?: string;
    endpoint?: string;
    authenticationType?: string;
    dependency?: string;
}
export type XSAppRouteProperties = keyof XSAppRoute;
export interface XSAppDocument {
    authenticationMethod?: string;
    routes?: XSAppRoute[];
    welcomeFile?: string;
}
export interface HTML5App {
    path: string;
    'build-parameters': {
        builder: string;
        'build-result': string;
        commands: string[];
        'supported-platforms': string[];
    };
    name: string;
    type: string;
}
