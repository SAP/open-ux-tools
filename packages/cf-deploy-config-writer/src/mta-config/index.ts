import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import { render } from 'ejs';
import { MtaConfig } from './mta';
import { getTemplatePath, setMtaDefaults, validateVersion } from '../utils';
import {
    MTAYamlFile,
    MTAVersion,
    MTADescription,
    deployMode,
    enableParallelDeployments,
    CDSAddMtaParams,
    CDSBinNotFound,
    CDSExecutable,
    MTABinNotFound,
    MTAExecutable,
    CDSXSUAAService,
    CDSDestinationService,
    CDSHTML5RepoService
} from '../constants';
import type { mta } from '@sap/mta-lib';
import { type MTABaseConfig, type CFBaseConfig, type CDSServiceType, RouterModuleType } from '../types';
import LoggerHelper from '../logger-helper';
import { sync } from 'hasbin';
import { spawnSync } from 'child_process';
import { t } from '../i18n';

/**
 * Get the MTA ID, read from the root path specified.
 *
 * @param rootPath Path to the root folder
 * @returns MTA ID if found
 */
export async function getMtaId(rootPath: string): Promise<string | undefined> {
    return (await getMtaConfig(rootPath))?.prefix;
}

/**
 *  Get the MTA configuration from the target folder.
 *
 * @param rootPath Path to the root folder
 * @returns MtaConfig instance if found
 */
export async function getMtaConfig(rootPath: string): Promise<MtaConfig | undefined> {
    let mtaConfig;
    for (let retries = 5; retries >= 0; retries--) {
        try {
            mtaConfig = await MtaConfig.newInstance(rootPath, LoggerHelper.logger);
            if (mtaConfig?.prefix) {
                break;
            }
        } catch (error) {
            await new Promise((resolve) => setTimeout(resolve, 300));
        }
    }
    LoggerHelper.logger?.info(`Read mta.yaml with prefix ${mtaConfig?.prefix}`);
    return mtaConfig;
}

/**
 *  Generate an MTA ID that is suitable for CF deployment.
 *
 * @param appId Name of the app, like `sap.ux.app` and restrict to 128 characters
 * @returns Name that's acceptable for mta.yaml
 */
export function toMtaModuleName(appId: string): string {
    return appId.replace(/[`~!@#$%^&*()_|+\-=?;:'",.<>]/gi, '').slice(0, 128);
}

/**
 * Create an MTA file in the target folder, needs to be written to disk as subsequent calls are dependent on it being on the file system i.e mta-lib.
 *
 * @param config writer configuration
 */
export function createMTA(config: MTABaseConfig): void {
    const mtaTemplate = readFileSync(getTemplatePath(`app/${MTAYamlFile}`), 'utf-8');
    const mtaContents = render(mtaTemplate, {
        id: `${config.mtaId.slice(0, 128)}`,
        mtaDescription: config.mtaDescription ?? MTADescription,
        mtaVersion: config.mtaVersion ?? MTAVersion
    });
    // Written to disk immediately! Subsequent calls are dependent on it being on the file system i.e mta-lib.
    writeFileSync(join(config.mtaPath, MTAYamlFile), mtaContents);
    LoggerHelper.logger?.debug(t('debug.mtaCreated', { mtaPath: config.mtaPath }));
}

/**
 *  Add the build parameters to the MTA configuration.
 *
 * @param mtaInstance MTA instance
 */
export async function addMtaBuildParams(mtaInstance: MtaConfig): Promise<void> {
    let params = await mtaInstance.getBuildParameters();
    params = { ...(params || {}), ...{} } as mta.ProjectBuildParameters;
    params['before-all'] ||= [];
    const buildParams: mta.BuildParameters = { builder: 'custom', commands: ['npm install'] };
    params['before-all'].push(buildParams);
    await mtaInstance.updateBuildParams(params);
}

/**
 * Add the deployment parameters to the MTA configuration.
 *
 * @param mtaInstance MTA instance
 */
export async function addMtaDeployParameters(mtaInstance: MtaConfig): Promise<void> {
    let params = await mtaInstance.getParameters();
    params = { ...(params || {}), ...{} } as mta.Parameters;
    params[deployMode] = 'html5-repo';
    params[enableParallelDeployments] = true;
    await mtaInstance.updateParameters(params);
}

/**
 * Validate MTA binary is available.
 *
 */
export function doesMTABinaryExist(): void {
    // CF Writer is dependent on the mta-lib library, which in turn relies on the mta executable being installed and available in the path
    if (!sync(MTAExecutable)) {
        throw new Error(MTABinNotFound);
    }
}

/**
 * Validate CDS binary is available.
 *
 */
export function doesCDSBinaryExist(): void {
    // CF Writer is dependent on the cds library
    if (!sync(CDSExecutable)) {
        throw new Error(CDSBinNotFound);
    }
}

/**
 * Generate an MTA using `cds` binary, appending any optional services passed in. Specific services are added if the router type is defined.
 *
 * @param cwd
 * @param options
 * @param routerType
 */
export function createCAPMTA(cwd: string, options?: CDSServiceType[], routerType?: RouterModuleType): void {
    let defaultOptions: CDSServiceType[] = [];
    if (routerType) {
        defaultOptions = [
            ...[CDSXSUAAService],
            ...(routerType === RouterModuleType.AppFront ? [] : [CDSDestinationService, CDSHTML5RepoService])
        ] as CDSServiceType[];
    }
    const cdsParams = [...CDSAddMtaParams, ...(options ?? []), ...defaultOptions];
    LoggerHelper.logger?.debug(t('debug.creatingMta', { cdsParams: cdsParams.toString() }));
    let result = spawnSync(CDSExecutable, cdsParams, { cwd });
    if (result?.error) {
        throw new Error(`Something went wrong creating mta.yaml! ${result.error}`);
    }
    // Ensure the package-lock is created otherwise mta build will fail
    const cmd = process.platform === 'win32' ? `npm.cmd` : 'npm';
    result = spawnSync(cmd, ['install', '--ignore-engines'], { cwd });
    if (result?.error) {
        throw new Error(`Something went wrong installing node modules! ${result.error}`);
    }
    LoggerHelper.logger?.debug(t('debug.capMtaCreated'));
}

/**
 * Validate the writer configuration to ensure all required parameters are present.
 *
 * @param config writer configuration
 */
export function validateMtaConfig(config: CFBaseConfig): void {
    // We use mta-lib, which in turn relies on the mta executable being installed and available in the path
    doesMTABinaryExist();

    if (!config.routerType || !config.mtaId || !config.mtaPath) {
        throw new Error(t('error.missingMtaParameters'));
    }
    if (config.mtaId.length > 128 || !/^[a-zA-Z_]/.test(config.mtaId)) {
        throw new Error(t('error.invalidMtaId'));
    }
    if (!/^[\w\-.]*$/.test(config.mtaId)) {
        throw new Error(t('error.invalidMtaIdWithChars'));
    }

    validateVersion(config.mtaVersion);

    if (
        config.abapServiceProvider &&
        (!config.abapServiceProvider.abapService || !config.abapServiceProvider.abapServiceName)
    ) {
        throw new Error(t('error.missingABAPServiceBindingDetails'));
    }

    setMtaDefaults(config);
}

export * from './mta';
