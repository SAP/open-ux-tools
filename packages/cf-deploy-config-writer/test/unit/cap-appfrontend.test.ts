import * as projectAccess from '@sap-ux/project-access';
import { join } from 'path';
import fsExtra from 'fs-extra';
import hasbin from 'hasbin';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { generateAppConfig, generateCAPConfig, RouterModuleType } from '../../src';
import type { Editor } from 'mem-fs-editor';
import { DefaultMTADestination } from '../../src/constants';
import { isAppStudio } from '@sap-ux/btp-utils';
import { NullTransport, ToolsLogger } from '@sap-ux/logger';
import { CommandRunner } from '@sap-ux/nodejs-utils';
import fs from 'fs';

jest.mock('@sap/mta-lib', () => {
    return {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        Mta: require('./mockMta').MockMta
    };
});

jest.mock('@sap-ux/btp-utils', () => ({
    ...jest.requireActual('@sap-ux/btp-utils'),
    isAppStudio: jest.fn()
}));
const isAppStudioMock = isAppStudio as jest.Mock;

let hasSyncMock: jest.SpyInstance;
let commandRunnerMock: jest.SpyInstance;
let unitTestFs: Editor;
const originalPlatform = process.platform;

describe('CF Writer with CAP App Frontend', () => {
    const outputDir = join(__dirname, '../test-output', 'capwithappfrontend');
    const logger = new ToolsLogger({
        transports: [new NullTransport()]
    });

    beforeEach(() => {
        jest.resetAllMocks();
        jest.restoreAllMocks();
        unitTestFs = create(createStorage());
        commandRunnerMock = jest.spyOn(CommandRunner.prototype, 'run').mockImplementation(() => ({ status: 0 } as any));
        isAppStudioMock.mockReturnValue(false);
        hasSyncMock = jest.spyOn(hasbin, 'sync').mockImplementation(() => true);
    });

    beforeAll(async () => {
        jest.clearAllMocks();
        jest.spyOn(hasbin, 'sync').mockReturnValue(true);
        fsExtra.removeSync(outputDir);
        jest.mock('hasbin', () => {
            return {
                ...(jest.requireActual('hasbin') as {}),
                sync: hasSyncMock
            };
        });
    });

    afterAll(async () => {
        jest.resetAllMocks();
        Object.defineProperty(process, 'platform', {
            value: originalPlatform
        });
    });

    describe('Generate deployment config', () => {
        test('Add HTML5 app to CAP App Frontend Project', async () => {
            const capPath = join(outputDir, 'cap');
            const getMtaPathMock = jest.spyOn(projectAccess, 'getMtaPath');
            const findCapProjectRootMock = jest.spyOn(projectAccess, 'findCapProjectRoot');
            fsExtra.mkdirSync(outputDir, { recursive: true });
            fsExtra.mkdirSync(capPath);
            fsExtra.copySync(join(__dirname, '../sample/capwithfrontend'), capPath);
            const localFs = await generateAppConfig(
                {
                    appPath: join(capPath, 'app/lrop'),
                    destinationName: DefaultMTADestination,
                    addAppFrontendRouter: true
                },
                unitTestFs
            );
            expect(getMtaPathMock).toBeCalledWith(expect.stringContaining(capPath));
            expect(findCapProjectRootMock).toHaveBeenCalledTimes(1);
            expect(findCapProjectRootMock).toBeCalledWith(expect.stringContaining(capPath));
            expect(commandRunnerMock).not.toHaveBeenCalled();
            expect(localFs.read(join(capPath, 'app/lrop', 'xs-app.json'))).toMatchSnapshot();
            expect(localFs.read(join(capPath, 'xs-security.json'))).toMatchSnapshot();
            expect(localFs.read(join(capPath, 'package.json'))).toMatchSnapshot();
            expect(fs.readFileSync(join(capPath, 'mta.yaml'), { encoding: 'utf8' })).toMatchSnapshot();
        });

        test('Generate CAP project with App Frontend Service', async () => {
            Object.defineProperty(process, 'platform', {
                value: 'win32'
            });
            const mtaId = 'base';
            const mtaPath = join(outputDir, mtaId);
            fsExtra.mkdirSync(mtaPath, { recursive: true });
            fsExtra.copySync(join(__dirname, `../sample/capcds`), mtaPath);
            const getCapProjectTypeMock = jest.spyOn(projectAccess, 'getCapProjectType').mockResolvedValue('CAPNodejs');
            const localFs = await generateCAPConfig(
                {
                    mtaPath,
                    mtaId,
                    routerType: RouterModuleType.AppFront
                },
                undefined,
                logger
            );
            expect(localFs.read(join(mtaPath, 'package.json'))).toMatchSnapshot();
            expect(fs.readFileSync(join(mtaPath, 'mta.yaml'), { encoding: 'utf8' })).toMatchSnapshot();
            expect(localFs.read(join(mtaPath, 'xs-security.json'))).toMatchSnapshot();
            expect(getCapProjectTypeMock).toHaveBeenCalled();
            expect(commandRunnerMock).not.toHaveBeenCalled();
        });
    });
});
