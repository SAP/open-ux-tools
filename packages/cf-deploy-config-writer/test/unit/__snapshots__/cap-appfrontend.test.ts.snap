// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CF Writer with CAP App Frontend Generate deployment config Add HTML5 app to CAP App Frontend Project 1`] = `
"{
  \\"welcomeFile\\": \\"/index.html\\",
  \\"authenticationMethod\\": \\"route\\",
  \\"routes\\": [
    {
      \\"source\\": \\"^/sap/(.*)$\\",
      \\"target\\": \\"/sap/$1\\",
      \\"destination\\": \\"cappapp-srv-api\\",
      \\"authenticationType\\": \\"ias\\",
      \\"csrfProtection\\": false
    },
    {
      \\"source\\": \\"^/resources/(.*)$\\",
      \\"target\\": \\"/resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    },
    {
      \\"source\\": \\"^/test-resources/(.*)$\\",
      \\"target\\": \\"/test-resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    },
    {
      \\"source\\": \\"^/logout-page.html$\\",
      \\"service\\": \\"app-front\\",
      \\"authenticationType\\": \\"none\\"
    },
    {
      \\"source\\": \\"^/index.html$\\",
      \\"service\\": \\"app-front\\",
      \\"cacheControl\\": \\"no-cache, no-store, must-revalidate\\"
    },
    {
      \\"source\\": \\"^(.*)$\\",
      \\"target\\": \\"$1\\",
      \\"service\\": \\"app-front\\",
      \\"authenticationType\\": \\"ias\\"
    }
  ]
}
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Add HTML5 app to CAP App Frontend Project 2`] = `
"{
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Add HTML5 app to CAP App Frontend Project 3`] = `
"{
    \\"dependencies\\": {
        \\"@sap/cds\\": \\"^8\\",
        \\"express\\": \\"^4\\",
        \\"@sap/xssec\\": \\"^4\\"
    },
    \\"devDependencies\\": {
        \\"@cap-js/cds-types\\": \\"^0.9.0\\",
        \\"@cap-js/sqlite\\": \\"^1\\",
        \\"@sap/cds-dk\\": \\"^8\\",
        \\"rimraf\\": \\"^5.0.5\\",
        \\"mbt\\": \\"^1.2.29\\"
    },
    \\"cds\\": {},
    \\"scripts\\": {
        \\"undeploy\\": \\"cf undeploy cappapp --delete-services --delete-service-keys --delete-service-brokers\\",
        \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
        \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\"
    }
}
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Add HTML5 app to CAP App Frontend Project 4`] = `
"_schema-version: '3.1'
ID: cappapp
version: 1.0.0
description: A simple CAP project.
parameters:
  enable-parallel-deployments: true
  deploy_mode: html5-repo
build-parameters:
  before-all:
    - builder: custom
      commands:
        - npm ci
        - npx -p @sap/cds-dk cds build --production
modules:
  - name: cappapp-srv
    type: nodejs
    path: gen/srv
    parameters:
      buildpack: nodejs_buildpack
    build-parameters:
      builder: npm-ci
    provides:
      - name: srv-api
        properties:
          srv-url: '\${default-url}'
    requires:
      - name: cappapp-auth
  - name: cappapp-app-content
    type: com.sap.application.content
    path: .
    requires:
      - name: cappapp-auth
      - name: cappapp-app-front
        parameters:
          content-target: true
    parameters:
      config:
        destinations:
          - name: ui5
            url: 'https://ui5.sap.com'
          - name: cappapp-srv-api
            url: '~{srv-api/srv-url}'
            forwardAuthToken: true
    build-parameters:
      build-result: resources
      requires:
        - name: comfioritoolslrop
          artifacts:
            - comfioritoolslrop.zip
          target-path: resources/
  - name: comfioritoolslrop
    type: html5
    path: app/lrop
    build-parameters:
      builder: custom
      build-result: dist
      commands:
        - npm install
        - 'npm run build:cf'
      supported-platforms: []
resources:
  - name: cappapp-auth
    type: org.cloudfoundry.managed-service
    parameters:
      service: xsuaa
      service-plan: application
      path: ./xs-security.json
      config:
        xsappname: 'cappapp-\${org}-\${space}'
        tenant-mode: dedicated
      service-name: cappapp-xsuaa-service
  - name: cappapp-app-front
    type: org.cloudfoundry.managed-service
    parameters:
      service: app-front
      service-name: cappapp-app-front-service
      service-plan: developer
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Generate CAP project with App Frontend Service 1`] = `
"{
  \\"name\\": \\"captestproject\\",
  \\"version\\": \\"1.0.0\\",
  \\"description\\": \\"A simple CAP project.\\",
  \\"repository\\": \\"<Add your repository here>\\",
  \\"license\\": \\"UNLICENSED\\",
  \\"private\\": true,
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"^8\\",
    \\"express\\": \\"^4\\",
    \\"@sap/xssec\\": \\"^4\\"
  },
  \\"devDependencies\\": {
    \\"@cap-js/cds-types\\": \\"^0.8.0\\",
    \\"@cap-js/sqlite\\": \\"^1\\",
    \\"@sap/cds-dk\\": \\"^8\\",
    \\"rimraf\\": \\"^5.0.5\\",
    \\"mbt\\": \\"^1.2.29\\"
  },
  \\"scripts\\": {
    \\"start\\": \\"cds-serve\\",
    \\"undeploy\\": \\"cf undeploy base --delete-services --delete-service-keys --delete-service-brokers\\",
    \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
    \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\"
  },
  \\"cds\\": {
    \\"requires\\": {
      \\"auth\\": \\"xsuaa\\",
      \\"connectivity\\": true,
      \\"destinations\\": true,
      \\"html5-repo\\": true
    }
  }
}
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Generate CAP project with App Frontend Service 2`] = `
"_schema-version: '3.2'
ID: base
description: Generated by Fiori Tools
version: 0.0.1
parameters:
  deploy_mode: html5-repo
  enable-parallel-deployments: true
build-parameters:
  before-all:
    - builder: custom
      commands:
        - npm install
        - npx cds build --production
        - cp -R db/data gen/srv/srv/
modules:
  - name: base-app-content
    type: com.sap.application.content
    path: .
    requires:
      - name: base-app-front
        parameters:
          content-target: true
      - name: base-uaa
    parameters:
      config:
        destinations:
          - name: ui5
            url: 'https://ui5.sap.com'
  - name: base-srv
    type: nodejs
    path: gen/srv
    parameters:
      buildpack: nodejs_buildpack
      readiness-health-check-type: http
      readiness-health-check-http-endpoint: /health
    build-parameters:
      builder: npm
      ignore:
        - node_modules/
    provides:
      - name: srv-api
        properties:
          srv-url: '\${default-url}'
    requires:
      - name: base-uaa
resources:
  - name: base-app-front
    type: org.cloudfoundry.managed-service
    parameters:
      service: app-front
      service-name: base-app-front-service
      service-plan: developer
  - name: base-uaa
    type: org.cloudfoundry.managed-service
    parameters:
      path: ./xs-security.json
      service: xsuaa
      service-name: base-xsuaa-service
      service-plan: application
      config:
        xsappname: 'base-\${org}-\${space}'
        tenant-mode: dedicated
"
`;

exports[`CF Writer with CAP App Frontend Generate deployment config Generate CAP project with App Frontend Service 3`] = `
"{
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
"
`;
